app/PaymentDrivers/CheckoutComPaymentDriver.php

```diff
@@ -356,27 +159,27 @@ public function createPayment($data, $status = Payment::STATUS_COMPLETED): Payme
     public function saveCard($state)
     {
         //some cards just can't be tokenized....
-        if(!$state['payment_response']->source['id'])
+        if (!$state['payment_response']->source['id'])
             return;
-        
-                    // [id] => src_hck5nsv3fljehbam2cvdm7fioa
-                    // [type] => card
-                    // [expiry_month] => 10
-                    // [expiry_year] => 2022
-                    // [scheme] => Visa
-                    // [last4] => 4242
-                    // [fingerprint] => 688192847DB9AE8A26C53776D036D5B8AD2CEAF1D5A8F5475F542B021041EFA1
-                    // [bin] => 424242
-                    // [card_type] => Credit
-                    // [card_category] => Consumer
-                    // [issuer] => JPMORGAN CHASE BANK NA
-                    // [issuer_country] => US
-                    // [product_id] => A
-                    // [product_type] => Visa Traditional
-                    // [avs_check] => S
-                    // [cvv_check] => Y
-                    // [payouts] => 1
-                    // [fast_funds] => d
+
+        // [id] => src_hck5nsv3fljehbam2cvdm7fioa
+        // [type] => card
+        // [expiry_month] => 10
+        // [expiry_year] => 2022
+        // [scheme] => Visa
+        // [last4] => 4242
+        // [fingerprint] => 688192847DB9AE8A26C53776D036D5B8AD2CEAF1D5A8F5475F542B021041EFA1
+        // [bin] => 424242
+        // [card_type] => Credit
+        // [card_category] => Consumer
+        // [issuer] => JPMORGAN CHASE BANK NA
+        // [issuer_country] => US
+        // [product_id] => A
+        // [product_type] => Visa Traditional
+        // [avs_check] => S
+        // [cvv_check] => Y
+        // [payouts] => 1
+        // [fast_funds] => d
 
         $payment_meta = new \stdClass;
```

Comment:
company_gateway_token creation should be moved into the BaseDriver. we should only have to pass just enough information from this class ie.... payment_meta, gateway_type_id, token... everything else can be resolved from the BaseDriver.

Reply:
Fixed here: https://github.com/invoiceninja/invoiceninja/pull/4202/commits/488b8d859df53ca6d74edae4b7e8a7dd0e722c9c#diff-9229767762db39ad236f5c2a30a13fa853f7f72e4eb1a072ae448ea7d4067493R257